{"cells":[{"cell_type":"markdown","id":"00274e26","metadata":{},"source":"## The following are the codes demonstrating the performance of SERM imputation for scRNA-seq data"},{"cell_type":"code","source":"# To illustrate the imputation calculation of SERM, the cellular taxonomy dataset analysed in the \r\n# manuscript will be automatically loaded. However, you can upload your own dataset, and impute the \r\n# data by SERM. Here the input data is saved as .mat file to reduce the data size (normally .csv file \r\n# needs more disk space). However, .csv files can also be loaded using python packages such as \"pandas\"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":4,"id":"f8e72a7a","metadata":{"collapsed":true},"outputs":[],"source":"# Import all the necessary Python packages\nimport os\nfrom functools import partial\nimport h5py\nimport matplotlib.pyplot as plt\nimport scipy\nimport scipy.io as sio\nfrom serm import *\nfrom sklearn.decomposition import PCA\nimport numpy as np\nfrom sklearn.manifold import TSNE\nimport magic\nimport random\n%matplotlib inline"},{"cell_type":"markdown","source":"Load data","metadata":{}},{"cell_type":"code","execution_count":5,"id":"aeab0cba","metadata":{},"outputs":[],"source":"# Load the original data (reference data), the downsampled data, and the data labels\n# The original data is from the work of Baryawno et al. \n# Reference: Baryawno, Ninib, et al. \"A cellular taxonomy \n# of the bone marrow stroma in homeostasis and leukemia.\" Cell 177.7 (2019): 1915-1932.\n# The data from the authors is very large with 23092 cells and 27998 genes. We choose 5% of the\n# top highly variable genes (1093 genes) to create a smaller dataset.\n# The following two code lines read the reference data described above\na=sio.loadmat('../data/data_CellTax_Original.mat')\ndataRef=a['dataRef']\n\n# The downsampled data (to 1%) were generated by using the technique described in Miao et al.\n# Please see the description of their codes at \n# https://www.bioconductor.org/packages/release/bioc/html/scRecover.html\n# We used their \"countsSampling\" function for the downsampling operation\n# Reference: Miao, Z., Li, J., & Zhang, X. (2019). scRecover: Discriminating true and \n# false zeros in single-cell RNA-seq data for imputation. bioRxiv, 665323.\n# The following two code lines read the downsampled data\na=sio.loadmat('../data/data_CellTax_Downsampled_1perc.mat')\ndataSample=a['dataSample']\n\n# Load data labels \na=sio.loadmat('../data/dataLabel_CellTax.mat')\ndataLabel=a['GT']"},{"cell_type":"markdown","source":"Set parameters of SERM and run it on the downsampled data","metadata":{}},{"cell_type":"code","source":"# Selection of the size of the region of interest (ROI)\r\n# We choose ROI size in such a way that the expression matrix\r\n# is divided into 8 (i.e. 4 by 2) ROIs. However, other ROI sizes are also\r\n# fine as SERM is generally very robust against the ROI size. \r\nROIsize = [6000,600]\r\n# Selection of percent of overlap between successive ROIs. The default value is 50%, but other \r\n# values can be chosen\r\npercOL = 0.5\r\n# Run SERM\r\nout_SERM = serm(dataSample,ROIsize,percOL,randomize=False) # If you want to randomize the rows of the data\r\n# before applying SERM, please choose randomize=True","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"For performance comparison, run MAGIC with default parameters","metadata":{}},{"cell_type":"code","source":"magic_operator = magic.MAGIC()\r\nout_MAGIC = magic_operator.fit_transform(dataSample)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Computation of Pearson coefficient for quantitative comparison","metadata":{}},{"cell_type":"code","source":"# Here, we compute the Pearson coefficient between the expression values of each gene of the cells from \r\n# the reference data and imputed data. Thus, for the loaded dataset with 1093 genes, we get 1093 \r\n# coefficients. We then use violin plot to show the  distribution of the coefficients\r\nsz=dataSample.shape\r\ncorrVecDownsampled=np.zeros(sz[1])\r\ncorrVecSERM=np.zeros(sz[1])\r\ncorrVecMAGIC=np.zeros(sz[1])\r\n\r\nfor i in range(sz[1]):\r\n    \r\n    downSampledGene=dataSample[:,i]\r\n    referenceGene=dataRef[:,i]\r\n    sermGene=out_SERM[:,i]\r\n    \r\n    a = np.corrcoef(downSampledGene, referenceGene)\r\n    corrVecDownsampled[i]=a[1,0]\r\n    \r\n    b= np.corrcoef(sermGene, referenceGene)\r\n    corrVecSERM[i]=b[1,0]\r\n    \r\n    magicGene=out_MAGIC[:,i]\r\n    b= np.corrcoef(magicGene, referenceGene)\r\n    corrVecMAGIC[i]=b[1,0]\r\n    \r\n# Stack all the coefficient values for plotting\r\ncorrVec=np.stack((corrVecDownsampled,corrVecMAGIC,corrVecSERM), axis=0)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Create violin plot of the Pearson coefficient values ","metadata":{}},{"cell_type":"code","source":"fig, (ax) = plt.subplots(nrows=1, ncols=1, sharey=True)\r\nlabels = ['Downsampled', 'MAGIC', 'SERM']\r\nplt.violinplot(corrVec.T,showmeans=True)\r\nplt.title('Comaprison of different imputation techniques')\r\nset_axis_style(ax, labels)\r\nax.set_ylabel('Pearson coefficient')","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"t-SNE visualization of the reference, the downsampled, and the imputed data","metadata":{}},{"cell_type":"code","source":"tsne_Original = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(dataRef)\r\nplt.figure(1)\r\nplt.scatter(tsne_Original[:,0],tsne_Original[:,1], c=dataLabel.flatten(),s=1)\r\nplt.xlabel(\"tSNE1\")\r\nplt.ylabel(\"tSNE2\")\r\nplt.title('t-SNE visualization of reference data')\r\n\r\ntsne_Downsampled = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(dataSample)\r\nplt.figure(2)\r\nplt.scatter(tsne_Downsampled[:,0],tsne_Downsampled[:,1], c=dataLabel.flatten(),s=1)\r\nplt.xlabel(\"tSNE1\")\r\nplt.ylabel(\"tSNE2\")\r\nplt.title('t-SNE visualization of downsampled data')\r\nplt.ylim(-30, 30)\r\n\r\ntsne_SERM = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(out_SERM)\r\nplt.figure(3)\r\nplt.scatter(tsne_SERM[:,0],tsne_SERM[:,1], c=dataLabel.flatten(),s=1)\r\nplt.xlabel(\"tSNE1\")\r\nplt.ylabel(\"tSNE2\")\r\nplt.title('t-SNE visualization of SERM-imputed data')\r\n\r\ntsne_MAGIC = TSNE(n_components=2, learning_rate='auto',init='random',random_state=1).fit_transform(out_MAGIC)\r\nplt.figure(4)\r\nplt.scatter(tsne_MAGIC[:,0],tsne_MAGIC[:,1], c=dataLabel.flatten(),s=1)\r\nplt.xlabel(\"tSNE1\")\r\nplt.ylabel(\"tSNE2\")\r\nplt.title('t-SNE visualization of MAGIC-imputed data')","metadata":{},"execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.8"}},"nbformat":4,"nbformat_minor":5}